File: desktop-app-vue/src/main/plugins/plugin-sandbox.js
/**
 * PluginSandbox - 插件沙箱
 *
 * 职责：
 * - 在隔离环境中执行插件代码
 * - 提供安全的global对象
 * - 超时控制和资源限制
 * - 生命周期钩子调用
 */

const { logger: pluginLogSink } = require("../utils/logger.js");
const { createPluginLogRedactor } = require("./plugin-log-redaction");
const {
  createPluginFailureDescriptor,
  createPluginMethodUnavailableError,
  createPluginOperationError,
} = require("./plugin-ipc-error-boundary");
const { isWithinDir } = require("../utils/path-boundary.js");
const vm = require("vm");
const fs = require("fs");
const EventEmitter = require("events");

const logger = createPluginLogRedactor(pluginLogSink, "PluginSandbox");
const MAX_SANDBOX_MODULES = 512;
const MAX_SANDBOX_MODULE_BYTES = 4 * 1024 * 1024;

function createSandboxModuleUnavailableError() {
  const error = new Error("Plugin sandbox module unavailable");
  error.code = "PLUGIN_SANDBOX_MODULE_UNAVAILABLE";
  return error;
}

// M2: _deps injection so tests can mock fs.promises (vi.mock cannot
// intercept fs.promises for inlined CJS modules)
const _deps = { fsp: fs.promises };

class PluginSandbox extends EventEmitter {
  constructor(pluginId, pluginPath, manifest, pluginAPI) {
    super();

    this.pluginId = pluginId;
    this.pluginPath = pluginPath;
    this.manifest = manifest;
    this.pluginAPI = pluginAPI;

    // 插件实例
    this.instance = null;

    // 执行上下文
    this.context = null;

    // 超时设置（毫秒）
    this.timeouts = {
      load: 10000, // 加载超时
      hook: 5000, // 钩子执行超时
      method: 30000, // 方法执行超时
    };

    // 状态
    this.state = "created"; // created, loaded, enabled, disabled, error

    // 插件通过沙箱 setTimeout/setInterval 创建的定时器句柄。destroy() 必须全部
    // 清除，否则插件的定时器回调会在沙箱销毁后继续运行（泄漏内存 + 销毁后仍执行
    // 插件代码）。
    this._activeTimers = new Set();
    this._moduleCache = new Map();
  }

  /**
   * 加载插件代码
   * @returns {Promise<Object>} 插件实例
   */
  async load() {
    logger.info(`[PluginSandbox] 加载插件: ${this.pluginId}`);

    try {
      this.state = "loading";

      // 1. 读取插件代码 (M2: 异步读取，避免启动期阻塞事件循环)
      const path = require("path");
      const entryFile = this.manifest.main || "index.js";
      const entryPath = path.join(this.pluginPath, entryFile);

      let code;
      try {
        code = await _deps.fsp.readFile(entryPath, "utf-8");
      } catch (err) {
        if (err.code === "ENOENT") {
          throw new Error(`插件入口文件不存在: ${entryPath}`);
        }
        logger.error("[PluginSandbox] 插件入口读取失败", err);
        throw createPluginOperationError("plugin");
      }

      // 2. 创建沙箱环境
      this.context = this.createSandboxContext();

      // 3. 在沙箱中执行代码
      const script = new vm.Script(
        `
        (function(module, exports, require, __dirname, __filename) {
          ${code}
        })
      `,
        {
          filename: entryPath,
          timeout: this.timeouts.load,
        },
      );

      // 模拟module系统
      const module = { exports: {} };
      const exports = module.exports;

      // 在VM上下文中执行
      const vmContext = vm.createContext(this.context);

      // 创建require函数。插件自有模块和依赖继续在同一VM上下文中执行，
      // 禁止退回宿主CommonJS loader。
      const customRequire = this.createRequireFunction(
        vmContext,
        path.dirname(entryPath),
      );

      try {
        const fn = script.runInContext(vmContext, {
          timeout: this.timeouts.load,
          displayErrors: true,
        });

        // 调用包装函数
        fn.call(
          exports,
          module,
          exports,
          customRequire,
          path.dirname(entryPath),
          entryPath,
        );
      } catch (error) {
        if (error.message.includes("timed out")) {
          throw new Error(`插件加载超时（${this.timeouts.load}ms）`);
        }
        logger.error("[PluginSandbox] 插件代码执行失败", error);
        throw createPluginOperationError("plugin");
      }

      // 4. 获取导出的插件类
      const PluginClass = module.exports;

      if (typeof PluginClass !== "function") {
        throw new Error("插件必须导出一个构造函数或类");
      }

      // 5. 实例化插件
      this.instance = new PluginClass();

      // 6. 验证插件接口
      this.validatePluginInterface();

      this.state = "loaded";
      this.emit("loaded", { pluginId: this.pluginId });

      logger.info(`[PluginSandbox] 插件加载成功: ${this.pluginId}`);

      return this.instance;
    } catch (error) {
      this.state = "error";
      this.emit("error", {
        pluginId: this.pluginId,
        ...createPluginFailureDescriptor("plugin"),
      });
      logger.error(`[PluginSandbox] 插件加载失败: ${this.pluginId}`, error);
      throw createPluginOperationError("plugin");
    }
  }

  /**
   * 创建沙箱上下文
   * @returns {Object} 沙箱上下文
   */
  createSandboxContext() {
    const forwardConsoleCall = (level, args) => {
      const sink = this.pluginAPI?.api?.utils?.[level];
      if (typeof sink !== "function") {
        return;
      }
      sink({
        event: "plugin-console",
        level,
        argumentCount: Math.min(args.length, 32),
        redacted: true,
      });
    };

    // 提供安全的全局对象
    const context = {
      // 标准JavaScript全局对象
      console: {
        log: (...args) => forwardConsoleCall("log", args),
        warn: (...args) => forwardConsoleCall("warn", args),
        error: (...args) => forwardConsoleCall("error", args),
      },

      // 定时器（带限制 + 句柄追踪，destroy() 时统一清理）
      setTimeout: (fn, delay) => {
        const id = setTimeout(
          () => {
            this._activeTimers.delete(id); // one-shot：触发后自动取消追踪
            try {
              fn();
            } catch (error) {
              logger.error(`[PluginSandbox] setTimeout错误:`, error);
            }
          },
          Math.min(delay, 60000),
        ); // 最大60秒
        this._activeTimers.add(id);
        return id;
      },

      setInterval: (fn, delay) => {
        const id = setInterval(
          () => {
            try {
              fn();
            } catch (error) {
              logger.error(`[PluginSandbox] setInterval错误:`, error);
            }
          },
          Math.max(delay, 100),
        ); // 最小100ms
        this._activeTimers.add(id);
        return id;
      },

      clearTimeout: (id) => {
        this._activeTimers.delete(id);
        clearTimeout(id);
      },
      clearInterval: (id) => {
        this._activeTimers.delete(id);
        clearInterval(id);
      },

      // Promise支持
      Promise: Promise,

      // 基本类型
      Object: Object,
      Array: Array,
      String: String,
      Number: Number,
      Boolean: Boolean,
      Date: Date,
      Math: Math,
      JSON: JSON,
      RegExp: RegExp,
      Error: Error,
      TypeError: TypeError,
      RangeError: RangeError,

      // Buffer（限制使用）
      Buffer: Buffer,

      // 插件API
      chainlesschain: this.pluginAPI.getAPI(),

      // 全局变量
      global: undefined, // 禁止访问global
      process: {
        env: {}, // 空的env对象
        version: process.version,
        platform: process.platform,
      },
    };

    return context;
  }

  /**
   * 创建受限的require函数
   * @returns {Function} require函数
   */
  createRequireFunction(
    vmContext,
    parentDirectory = this.pluginPath,
    moduleCache = this._moduleCache,
  ) {
    // 允许的内置模块白名单
    const allowedModules = [
      "crypto",
      "path",
      "url",
      "querystring",
      "util",
      "events",
    ];

    return (moduleName) => {
      // 检查是否为允许的内置模块
      if (allowedModules.includes(moduleName)) {
        return require(moduleName);
      }

      if (
        typeof moduleName !== "string" ||
        !vmContext ||
        typeof vmContext !== "object"
      ) {
        throw createSandboxModuleUnavailableError();
      }

      const path = require("path");
      let resolvedPath;
      try {
        if (moduleName.startsWith(".") || path.isAbsolute(moduleName)) {
          resolvedPath = require.resolve(
            path.resolve(parentDirectory, moduleName),
          );
        } else {
          resolvedPath = require.resolve(moduleName, {
            paths: [parentDirectory, this.pluginPath],
          });
        }
      } catch (_error) {
        throw createSandboxModuleUnavailableError();
      }

      let pluginRoot;
      let moduleRealPath;
      try {
        pluginRoot = fs.realpathSync(this.pluginPath);
        moduleRealPath = fs.realpathSync(resolvedPath);
      } catch (_error) {
        throw createSandboxModuleUnavailableError();
      }

      if (!isWithinDir(pluginRoot, moduleRealPath)) {
        throw createSandboxModuleUnavailableError();
      }

      if (moduleCache.has(moduleRealPath)) {
        return moduleCache.get(moduleRealPath).exports;
      }
      if (moduleCache.size >= MAX_SANDBOX_MODULES) {
        throw createSandboxModuleUnavailableError();
      }

      let source;
      try {
        const stats = fs.statSync(moduleRealPath);
        if (
          !stats.isFile() ||
          stats.size < 0 ||
          stats.size > MAX_SANDBOX_MODULE_BYTES
        ) {
          throw createSandboxModuleUnavailableError();
        }
        source = fs.readFileSync(moduleRealPath, "utf8");
      } catch (_error) {
        throw createSandboxModuleUnavailableError();
      }

      const extension = path.extname(moduleRealPath).toLowerCase();
      if (extension === ".json") {
        try {
          const record = { exports: JSON.parse(source) };
          moduleCache.set(moduleRealPath, record);
          return record.exports;
        } catch (_error) {
          throw createSandboxModuleUnavailableError();
        }
      }
      if (extension !== ".js" && extension !== ".cjs") {
        throw createSandboxModuleUnavailableError();
      }

      const record = { exports: {} };
      moduleCache.set(moduleRealPath, record);
      try {
        const factory = new vm.Script(
          `(function(module, exports, require, __dirname, __filename) {\n${source}\n})`,
          {
            filename: moduleRealPath,
            displayErrors: false,
          },
        ).runInContext(vmContext, {
          timeout: this.timeouts.load,
          displayErrors: false,
        });
        const localRequire = this.createRequireFunction(
          vmContext,
          path.dirname(moduleRealPath),
          moduleCache,
        );
        factory.call(
          record.exports,
          record,
          record.exports,
          localRequire,
          path.dirname(moduleRealPath),
          moduleRealPath,
        );
        return record.exports;
      } catch (_error) {
        moduleCache.delete(moduleRealPath);
        throw createSandboxModuleUnavailableError();
      }
    };
  }

  /**
   * 验证插件接口
   */
  validatePluginInterface() {
    if (!this.instance) {
      throw new Error("插件实例不存在");
    }

    // 检查必需的方法（可选）
    const optionalMethods = ["onEnable", "onDisable", "onLoad", "onUnload"];

    optionalMethods.forEach((method) => {
      if (
        this.instance[method] &&
        typeof this.instance[method] !== "function"
      ) {
        logger.warn(`[PluginSandbox] ${method} 应该是一个函数`);
      }
    });
  }

  /**
   * 调用插件钩子
   * @param {string} hookName - 钩子名称
   * @param {...any} args - 参数
   * @returns {Promise<any>} 钩子返回值
   */
  async callHook(hookName, ...args) {
    if (!this.instance) {
      throw new Error("插件未加载");
    }

    const hook = this.instance[hookName];

    if (!hook || typeof hook !== "function") {
      logger.info(`[PluginSandbox] 插件没有 ${hookName} 钩子，跳过`);
      return null;
    }

    logger.info(`[PluginSandbox] 调用钩子: ${this.pluginId}.${hookName}`);

    try {
      // 使用Promise.race实现超时控制
      const timeoutPromise = new Promise((_, reject) => {
        setTimeout(() => {
          reject(
            new Error(`钩子 ${hookName} 执行超时（${this.timeouts.hook}ms）`),
          );
        }, this.timeouts.hook);
      });

      const hookPromise = Promise.resolve(hook.call(this.instance, ...args));

      const result = await Promise.race([hookPromise, timeoutPromise]);

      logger.info(`[PluginSandbox] 钩子执行成功: ${this.pluginId}.${hookName}`);

      return result;
    } catch (error) {
      logger.error(
        `[PluginSandbox] 钩子执行失败: ${this.pluginId}.${hookName}`,
        error,
      );
      this.emit("hook-error", {
        pluginId: this.pluginId,
        hookName,
        ...createPluginFailureDescriptor("plugin"),
      });
      throw createPluginOperationError("plugin");
    }
  }

  /**
   * 启用插件
   * @returns {Promise<void>}
   */
  async enable() {
    if (this.state === "enabled") {
      logger.info(`[PluginSandbox] 插件已启用: ${this.pluginId}`);
      return;
    }

    try {
      await this.callHook("onEnable");
      this.state = "enabled";
      this.emit("enabled", { pluginId: this.pluginId });
      logger.info(`[PluginSandbox] 插件已启用: ${this.pluginId}`);
    } catch (_error) {
      this.state = "error";
      throw createPluginOperationError("plugin");
    }
  }

  /**
   * 禁用插件
   * @returns {Promise<void>}
   */
  async disable() {
    if (this.state === "disabled") {
      logger.info(`[PluginSandbox] 插件已禁用: ${this.pluginId}`);
      return;
    }

    try {
      await this.callHook("onDisable");
      this.state = "disabled";
      this.emit("disabled", { pluginId: this.pluginId });
      logger.info(`[PluginSandbox] 插件已禁用: ${this.pluginId}`);
    } catch (_error) {
      this.state = "error";
      throw createPluginOperationError("plugin");
    }
  }

  /**
   * 卸载插件
   * @returns {Promise<void>}
   */
  async unload() {
    try {
      await this.callHook("onUnload");
      this.instance = null;
      this.context = null;
      this.state = "unloaded";
      this.emit("unloaded", { pluginId: this.pluginId });
      logger.info(`[PluginSandbox] 插件已卸载: ${this.pluginId}`);
    } catch (_error) {
      this.state = "error";
      throw createPluginOperationError("plugin");
    }
  }

  /**
   * 调用插件方法
   * @param {string} methodName - 方法名
   * @param {...any} args - 参数
   * @returns {Promise<any>} 方法返回值
   */
  async callMethod(methodName, ...args) {
    if (!this.instance) {
      throw new Error("插件未加载");
    }

    const method = this.instance[methodName];

    if (!method || typeof method !== "function") {
      throw createPluginMethodUnavailableError();
    }

    logger.info(`[PluginSandbox] 调用方法: ${this.pluginId}.${methodName}`);

    try {
      // 超时控制
      const timeoutPromise = new Promise((_, reject) => {
        setTimeout(() => {
          reject(
            new Error(
              `方法 ${methodName} 执行超时（${this.timeouts.method}ms）`,
            ),
          );
        }, this.timeouts.method);
      });

      const methodPromise = Promise.resolve(
        method.call(this.instance, ...args),
      );

      const result = await Promise.race([methodPromise, timeoutPromise]);

      logger.info(
        `[PluginSandbox] 方法执行成功: ${this.pluginId}.${methodName}`,
      );

      return result;
    } catch (error) {
      logger.error(
        `[PluginSandbox] 方法执行失败: ${this.pluginId}.${methodName}`,
        error,
      );
      this.emit("method-error", {
        pluginId: this.pluginId,
        methodName,
        ...createPluginFailureDescriptor("plugin"),
      });
      throw createPluginOperationError("plugin");
    }
  }

  /**
   * 获取插件状态
   * @returns {string} 状态
   */
  getState() {
    return this.state;
  }

  /**
   * 获取插件实例
   * @returns {Object|null} 插件实例
   */
  getInstance() {
    return this.instance;
  }

  /**
   * 销毁沙箱
   */
  destroy() {
    // 清除插件残留的定时器，避免销毁后插件回调继续运行 / 泄漏。
    for (const id of this._activeTimers) {
      clearTimeout(id);
      clearInterval(id); // Node 的 Timeout 句柄可被任一方法清除，双清安全
    }
    this._activeTimers.clear();
    this._moduleCache.clear();

    this.removeAllListeners();
    this.instance = null;
    this.context = null;
    this.state = "destroyed";
  }
}

module.exports = PluginSandbox;
module.exports._deps = _deps;

File: ios-app/ChainlessChain/Features/Plugins/Services/PluginSandbox.swift
import Foundation
import JavaScriptCore

// MARK: - PluginSandbox
/// Plugin sandbox for isolated execution
/// Ported from PC: plugins/plugin-sandbox.js
///
/// Features:
/// - Isolated JavaScript execution via JavaScriptCore
/// - Secure global object with limited access
/// - Timeout control for operations
/// - Lifecycle hook management
///
/// Version: v1.7.0
/// Date: 2026-02-10

// MARK: - Sandbox Errors

enum PluginSandboxError: Error, LocalizedError {
    case notLoaded
    case loadTimeout
    case hookTimeout(String)
    case methodTimeout(String)
    case invalidPluginExport
    case methodNotFound(String)
    case executionFailed(String)
    case securityViolation(String)

    var errorDescription: String? {
        switch self {
        case .notLoaded:
            return "Plugin not loaded"
        case .loadTimeout:
            return "Plugin load timed out"
        case .hookTimeout(let hook):
            return "Hook '\(hook)' execution timed out"
        case .methodTimeout(let method):
            return "Method '\(method)' execution timed out"
        case .invalidPluginExport:
            return "Plugin must export a valid constructor or object"
        case .methodNotFound(let method):
            return "Plugin method not found: \(method)"
        case .executionFailed(let reason):
            return "Execution failed: \(reason)"
        case .securityViolation(let reason):
            return "Security violation: \(reason)"
        }
    }
}

// MARK: - Sandbox Delegate

protocol PluginSandboxDelegate: AnyObject {
    func sandbox(_ sandbox: PluginSandbox, didLoad pluginId: String)
    func sandbox(_ sandbox: PluginSandbox, didEnable pluginId: String)
    func sandbox(_ sandbox: PluginSandbox, didDisable pluginId: String)
    func sandbox(_ sandbox: PluginSandbox, didUnload pluginId: String)
    func sandbox(_ sandbox: PluginSandbox, didFailWithError error: Error, pluginId: String)
    func sandbox(_ sandbox: PluginSandbox, didLogMessage message: String, level: String, pluginId: String)
}

// MARK: - Plugin Sandbox

/// Sandbox environment for plugin execution
@MainActor
class PluginSandbox: ObservableObject {

    // MARK: - Properties

    let pluginId: String
    let pluginPath: URL
    let manifest: PluginManifest

    private let logger = Logger.shared
    private var jsContext: JSContext?
    private var pluginInstance: JSValue?

    weak var delegate: PluginSandboxDelegate?

    /// Timeout settings (milliseconds)
    struct Timeouts {
        var load: TimeInterval = 10.0
        var hook: TimeInterval = 5.0
        var method: TimeInterval = 30.0
    }
    var timeouts = Timeouts()

    /// Current state
    @Published private(set) var state: PluginState = .created

    /// Plugin API provider
    private var apiProvider: PluginAPIProvider?

    // MARK: - Initialization

    init(pluginId: String, pluginPath: URL, manifest: PluginManifest, apiProvider: PluginAPIProvider?) {
        self.pluginId = pluginId
        self.pluginPath = pluginPath
        self.manifest = manifest
        self.apiProvider = apiProvider
    }

    // MARK: - Loading

    /// Load plugin code in sandbox
    func load() async throws {
        logger.info("[PluginSandbox] Loading plugin: \(pluginId)")
        state = .loading

        do {
            // Create JavaScript context
            jsContext = JSContext()
            guard let context = jsContext else {
                throw PluginSandboxError.executionFailed("Failed to create JS context")
            }

            // Set up sandbox environment
            setupSandboxEnvironment(context)

            // Read plugin code
            let entryFile = manifest.main
            let entryPath = pluginPath.appendingPathComponent(entryFile)

            guard FileManager.default.fileExists(atPath: entryPath.path) else {
                throw PluginSandboxError.executionFailed("Entry file not found: \(entryPath.path)")
            }

            let code = try String(contentsOf: entryPath, encoding: .utf8)

            // Execute plugin code with timeout
            let result = try await withTimeout(timeouts.load) {
                return try self.executePluginCode(code, in: context)
            }

            // Validate plugin export
            guard let instance = result, !instance.isUndefined else {
                throw PluginSandboxError.invalidPluginExport
            }

            pluginInstance = instance

            // Validate plugin interface
            validatePluginInterface()

            state = .loaded
            delegate?.sandbox(self, didLoad: pluginId)
            logger.info("[PluginSandbox] Plugin loaded successfully: \(pluginId)")

        } catch {
            state = .error
            delegate?.sandbox(self, didFailWithError: error, pluginId: pluginId)
            logger.error("[PluginSandbox] Plugin load failed: \(pluginId) - \(error)")
            throw error
        }
    }

    /// Set up sandbox environment with limited global access
    private func setupSandboxEnvironment(_ context: JSContext) {
        // Exception handler
        context.exceptionHandler = { [weak self] ctx, exception in
            guard let self = self else { return }
            let message = exception?.toString() ?? "Unknown error"
            self.logger.error("[PluginSandbox] JS Exception: \(message)")
            self.delegate?.sandbox(self, didLogMessage: message, level: "error", pluginId: self.pluginId)
        }

        // Console logging
        let consoleLog: @convention(block) (String) -> Void = { [weak self] message in
            guard let self = self else { return }
            self.logger.info("[Plugin:\(self.pluginId)] \(message)")
            self.delegate?.sandbox(self, didLogMessage: message, level: "info", pluginId: self.pluginId)
        }

        let consoleWarn: @convention(block) (String) -> Void = { [weak self] message in
            guard let self = self else { return }
            self.logger.warning("[Plugin:\(self.pluginId)] \(message)")
            self.delegate?.sandbox(self, didLogMessage: message, level: "warn", pluginId: self.pluginId)
        }

        let consoleError: @convention(block) (String) -> Void = { [weak self] message in
            guard let self = self else { return }
            self.logger.error("[Plugin:\(self.pluginId)] \(message)")
            self.delegate?.sandbox(self, didLogMessage: message, level: "error", pluginId: self.pluginId)
        }

        // Create console object
        let console = JSValue(newObjectIn: context)
        console?.setObject(consoleLog, forKeyedSubscript: "log" as NSString)
        console?.setObject(consoleWarn, forKeyedSubscript: "warn" as NSString)
        console?.setObject(consoleError, forKeyedSubscript: "error" as NSString)
        context.setObject(console, forKeyedSubscript: "console" as NSString)

        // setTimeout (limited to 60 seconds)
        let setTimeout: @convention(block) (JSValue, Double) -> Int32 = { callback, delay in
            let limitedDelay = min(delay, 60000)
            let id = Int32.random(in: 1...Int32.max)

            DispatchQueue.main.asyncAfter(deadline: .now() + limitedDelay / 1000) {
                callback.call(withArguments: [])
            }

            return id
        }
        context.setObject(setTimeout, forKeyedSubscript: "setTimeout" as NSString)

        // setInterval (minimum 100ms)
        let setInterval: @convention(block) (JSValue, Double) -> Int32 = { callback, delay in
            let limitedDelay = max(delay, 100)
            let id = Int32.random(in: 1...Int32.max)

            Timer.scheduledTimer(withTimeInterval: limitedDelay / 1000, repeats: true) { _ in
                callback.call(withArguments: [])
            }

            return id
        }
        context.setObject(setInterval, forKeyedSubscript: "setInterval" as NSString)

        // Plugin API (chainlesschain global)
        if let apiProvider = apiProvider {
            let api = apiProvider.buildAPI(for: pluginId, manifest: manifest)
            context.setObject(api, forKeyedSubscript: "chainlesschain" as NSString)
        }

        // Module system
        let moduleExports = JSValue(newObjectIn: context)
        let module = JSValue(newObjectIn: context)
        module?.setObject(moduleExports, forKeyedSubscript: "exports" as NSString)
        context.setObject(module, forKeyedSubscript: "module" as NSString)
        context.setObject(moduleExports, forKeyedSubscript: "exports" as NSString)

        // Create require function
        let require: @convention(block) (String) -> JSValue? = { [weak self] moduleName in
            return self?.handleRequire(moduleName, context: context)
        }
        context.setObject(require, forKeyedSubscript: "require" as NSString)

        // __dirname and __filename
        context.setObject(pluginPath.path, forKeyedSubscript: "__dirname" as NSString)
        context.setObject(pluginPath.appendingPathComponent(manifest.main).path, forKeyedSubscript: "__filename" as NSString)
    }

    /// Handle require calls with security restrictions
    private func handleRequire(_ moduleName: String, context: JSContext) -> JSValue? {
        // Allowed built-in modules
        let allowedModules = ["crypto", "path", "url", "events", "util"]

        if allowedModules.contains(moduleName) {
            // Return mock implementations for allowed modules
            return createMockModule(moduleName, context: context)
        }

        // Relative paths within plugin directory
        if moduleName.hasPrefix(".") || moduleName.hasPrefix("/") {
            let resolvedPath = pluginPath.appendingPathComponent(moduleName)

            // Security check: must be within plugin directory
            guard resolvedPath.path.hasPrefix(pluginPath.path) else {
                logger.error("[PluginSandbox] Security: Blocked require outside plugin directory: \(moduleName)")
                return nil
            }

            // Load the file
            if let code = try? String(contentsOf: resolvedPath, encoding: .utf8) {
                context.evaluateScript(code)
                return context.objectForKeyedSubscript("module")?.objectForKeyedSubscript("exports")
            }
        }

        logger.warning("[PluginSandbox] Blocked require: \(moduleName)")
        return nil
    }

    /// Create mock implementations for allowed modules
    private func createMockModule(_ name: String, context: JSContext) -> JSValue? {
        let mock = JSValue(newObjectIn: context)

        switch name {
        case "crypto":
            // Basic crypto functions
            let randomUUID: @convention(block) () -> String = {
                UUID().uuidString
            }
            mock?.setObject(randomUUID, forKeyedSubscript: "randomUUID" as NSString)

        case "path":
            // Path utilities
            let join: @convention(block) ([String]) -> String = { parts in
                parts.joined(separator: "/")
            }
            mock?.setObject(join, forKeyedSubscript: "join" as NSString)

            let basename: @convention(block) (String) -> String = { path in
                URL(fileURLWithPath: path).lastPathComponent
            }
            mock?.setObject(basename, forKeyedSubscript: "basename" as NSString)

        case "events":
            // Basic EventEmitter
            let EventEmitter: @convention(block) () -> JSValue? = { [weak context] in
                guard let ctx = context else { return nil }
                let emitter = JSValue(newObjectIn: ctx)
                let listeners = JSValue(newObjectIn: ctx)
                emitter?.setObject(listeners, forKeyedSubscript: "_listeners" as NSString)
                return emitter
            }
            mock?.setObject(EventEmitter, forKeyedSubscript: "EventEmitter" as NSString)

        default:
            break
        }

        return mock
    }

    /// Execute plugin code and return the export
    private func executePluginCode(_ code: String, in context: JSContext) throws -> JSValue? {
        // Wrap code in IIFE to get exports
        let wrappedCode = """
        (function() {
            \(code)
            return module.exports;
        })()
        """

        let result = context.evaluateScript(wrappedCode)

        // Check for exceptions
        if let exception = context.exception {
            throw PluginSandboxError.executionFailed(exception.toString())
        }

        // If result is a constructor, instantiate it
        if let constructor = result, constructor.isObject {
            // Check if it's a class/constructor
            if let instance = constructor.construct(withArguments: []) {
                return instance
            }
            return constructor
        }

        return result
    }

    /// Validate plugin implements expected interface
    private func validatePluginInterface() {
        guard let instance = pluginInstance else { return }

        let optionalMethods = ["onLoad", "onUnload", "onEnable", "onDisable"]

        for method in optionalMethods {
            if let methodValue = instance.objectForKeyedSubscript(method),
               !methodValue.isUndefined && !methodValue.isNull {
                // Method exists - verify it's a function
                if !methodValue.isObject {
                    logger.warning("[PluginSandbox] \(method) should be a function")
                }
            }
        }
    }

    // MARK: - Lifecycle Hooks

    /// Call plugin hook with timeout
    func callHook(_ hookName: String, arguments: [Any] = []) async throws -> Any? {
        guard let instance = pluginInstance else {
            throw PluginSandboxError.notLoaded
        }

        guard let hook = instance.objectForKeyedSubscript(hookName),
              !hook.isUndefined && !hook.isNull else {
            logger.info("[PluginSandbox] Plugin has no \(hookName) hook, skipping")
            return nil
        }

        logger.info("[PluginSandbox] Calling hook: \(pluginId).\(hookName)")

        do {
            let result = try await withTimeout(timeouts.hook) {
                return hook.call(withArguments: arguments)
            }

            logger.info("[PluginSandbox] Hook executed successfully: \(pluginId).\(hookName)")
            return result?.toObject()

        } catch {
            logger.error("[PluginSandbox] Hook execution failed: \(pluginId).\(hookName) - \(error)")
            throw PluginSandboxError.hookTimeout(hookName)
        }
    }

    /// Enable plugin
    func enable() async throws {
        guard state != .enabled else {
            logger.info("[PluginSandbox] Plugin already enabled: \(pluginId)")
            return
        }

        state = .enabling

        do {
            _ = try await callHook("onEnable")
            state = .enabled
            delegate?.sandbox(self, didEnable: pluginId)
            logger.info("[PluginSandbox] Plugin enabled: \(pluginId)")
        } catch {
            state = .error
            throw error
        }
    }

    /// Disable plugin
    func disable() async throws {
        guard state == .enabled else {
            logger.info("[PluginSandbox] Plugin not enabled: \(pluginId)")
            return
        }

        do {
            _ = try await callHook("onDisable")
            state = .disabled
            delegate?.sandbox(self, didDisable: pluginId)
            logger.info("[PluginSandbox] Plugin disabled: \(pluginId)")
        } catch {
            state = .error
            throw error
        }
    }

    /// Unload plugin
    func unload() async throws {
        do {
            _ = try await callHook("onUnload")
            pluginInstance = nil
            jsContext = nil
            state = .unloaded
            delegate?.sandbox(self, didUnload: pluginId)
            logger.info("[PluginSandbox] Plugin unloaded: \(pluginId)")
        } catch {
            state = .error
            throw error
        }
    }

    // MARK: - Method Calls

    /// Call plugin method
    func callMethod(_ methodName: String, arguments: [Any] = []) async throws -> Any? {
        guard let instance = pluginInstance else {
            throw PluginSandboxError.notLoaded
        }

        guard let method = instance.objectForKeyedSubscript(methodName),
              !method.isUndefined && !method.isNull else {
            throw PluginSandboxError.methodNotFound(methodName)
        }

        logger.info("[PluginSandbox] Calling method: \(pluginId).\(methodName)")

        do {
            let result = try await withTimeout(timeouts.method) {
                return method.call(withArguments: arguments)
            }

            logger.info("[PluginSandbox] Method executed successfully: \(pluginId).\(methodName)")
            return result?.toObject()

        } catch {
            logger.error("[PluginSandbox] Method execution failed: \(pluginId).\(methodName) - \(error)")
            throw PluginSandboxError.methodTimeout(methodName)
        }
    }

    // MARK: - Getters

    /// Get plugin instance
    func getInstance() -> JSValue? {
        return pluginInstance
    }

    /// Get current state
    func getState() -> PluginState {
        return state
    }

    // MARK: - Cleanup

    /// Destroy sandbox
    func destroy() {
        pluginInstance = nil
        jsContext = nil
        state = .destroyed
        logger.info("[PluginSandbox] Sandbox destroyed: \(pluginId)")
    }

    // MARK: - Helpers

    /// Execute with timeout
    private func withTimeout<T>(_ timeout: TimeInterval, operation: @escaping () throws -> T) async throws -> T {
        try await withThrowingTaskGroup(of: T.self) { group in
            group.addTask {
                try operation()
            }

            group.addTask {
                try await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
                throw PluginSandboxError.loadTimeout
            }

            guard let result = try await group.next() else {
                throw PluginSandboxError.loadTimeout
            }

            group.cancelAll()
            return result
        }
    }
}

// MARK: - Plugin API Provider Protocol

/// Protocol for providing plugin APIs
protocol PluginAPIProvider: AnyObject {
    func buildAPI(for pluginId: String, manifest: PluginManifest) -> [String: Any]
    func checkPermission(_ permission: PluginPermission, for pluginId: String) -> Bool
}
